File: app/src/main/java/ru/michanic/mymot/UI/Adapters/ModelsExpandableListAdapter.java
package ru.michanic.mymot.UI.Adapters;

import android.content.Context;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import ru.michanic.mymot.Enums.CellAccessoryType;
import ru.michanic.mymot.Models.Category;
import ru.michanic.mymot.Models.FilterModelItem;
import ru.michanic.mymot.Models.Location;
import ru.michanic.mymot.Models.Manufacturer;
import ru.michanic.mymot.Models.Model;
import ru.michanic.mymot.Models.SectionModelItem;
import ru.michanic.mymot.Protocols.Const;
import ru.michanic.mymot.R;
import ru.michanic.mymot.UI.Cells.SimpleCell;
import ru.michanic.mymot.Utils.DataManager;

public class ModelsExpandableListAdapter extends BaseExpandableListAdapter {

    public Context context;
    private List<FilterModelItem> topCells;

    public ModelsExpandableListAdapter(Context context) {
        this.context = context;

        topCells = new ArrayList<>();
        topCells.add(new FilterModelItem("Все мотоциклы", false));

        DataManager dataManager = new DataManager();

        for (Manufacturer manufacturer: dataManager.getManufacturers(true)) {
            topCells.add(new FilterModelItem(manufacturer.getName()));
            topCells.add(new FilterModelItem("Все мотоциклы " + manufacturer.getName(), false));

            for (Category category: dataManager.getCategories(true)) {
                List<Model> models = dataManager.getManufacturerModels(manufacturer, category);
                if (models.size() > 0) {
                    topCells.add(new FilterModelItem(category, models));
                }
            }
        }

    }

    @Override
    public int getGroupCount() {
        return topCells.size();
    }

    @Override
    public int getChildrenCount(int groupPosition) {
        FilterModelItem item = topCells.get(groupPosition);
        switch (item.type) {
            case FilterModelItem.SECTION_TITLE:
                return 0;
            case FilterModelItem.SIMPLE_CELL:
                return 0;
            case FilterModelItem.CATEGORY_CELL:
                return item.getModels().size();
        }
        return 0;
    }

    @Override
    public Object getGroup(int groupPosition) {
        return groupPosition;
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        return childPosition;
    }

    @Override
    public long getGroupId(int groupPosition) {
        return groupPosition;
    }

    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return childPosition;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public View getGroupView(int groupPosition, boolean isExpanded, View convertView, ViewGroup parent) {
        FilterModelItem item = topCells.get(groupPosition);
        switch (item.type) {
            case FilterModelItem.SECTION_TITLE:
                View sectionView = View.inflate(context, R.layout.cell_section_title, null);
                TextView title = (TextView) sectionView.findViewById(R.id.section_title);
                title.setText(item.getTitle());
                return sectionView;
            case FilterModelItem.SIMPLE_CELL:
                View view = View.inflate(context, R.layout.cell_simple, null);
                SimpleCell.fillWithTitle(view, item.getTitle(), CellAccessoryType.HIDDEN, 1);
                return view;
            case FilterModelItem.CATEGORY_CELL:
                View categoryView = View.inflate(context, R.layout.cell_simple, null);
                SimpleCell.fillWithTitle(categoryView, item.getTitle(), isExpanded ? CellAccessoryType.TOP : CellAccessoryType.BOTTOM, 1);
                return categoryView;
        }
        return null;
    }

    @Override
    public View getChildView(int groupPosition, int childPosition, boolean isLastChild, View convertView, ViewGroup parent) {
        View view = View.inflate(context, R.layout.cell_models_list, null);

        ImageView imageView = (ImageView) view.findViewById(R.id.cell_image);
        TextView modelTitle = (TextView) view.findViewById(R.id.model_title);
        TextView years = (TextView) view.findViewById(R.id.model_years);

        Model model = topCells.get(groupPosition).getModels().get(childPosition);
        Picasso.get().load(Const.DOMAIN + model.getPreview_picture()).placeholder(R.drawable.ic_placeholder).into(imageView);
        modelTitle.setText(model.getName());
        years.setText(model.getYears());

        return view;
    }

    @Override
    public boolean isChildSelectable(int groupPosition, int childPosition) {
        return true;
    }
}

File: app/src/main/java/ru/michanic/mymot/Kotlin/UI/Adapters/ModelsExpandableListAdapter.kt
package ru.michanic.mymot.Kotlin.UI.Adapters

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.ImageView
import android.widget.TextView
import com.squareup.picasso.Picasso
import ru.michanic.mymot.Enums.CellAccessoryType
import ru.michanic.mymot.Kotlin.Models.FilterModelItem
import ru.michanic.mymot.MyMotApplication
import ru.michanic.mymot.Protocols.Const
import ru.michanic.mymot.R
import ru.michanic.mymot.Kotlin.UI.Cells.SimpleCell

class ModelsExpandableListAdapter(
    var context: Context,
    private val topCells: List<FilterModelItem>
) : BaseExpandableListAdapter() {
    override fun getGroupCount(): Int {
        return topCells.size
    }

    override fun getChildrenCount(groupPosition: Int): Int {
        val item = topCells[groupPosition]
        when (item.type) {
            FilterModelItem.SECTION_TITLE -> return 0
            FilterModelItem.SIMPLE_CELL -> return 0
            FilterModelItem.CATEGORY_CELL -> return item.models?.size ?: 0
        }
        return 0
    }

    override fun getGroup(groupPosition: Int): Any {
        return groupPosition
    }

    override fun getChild(groupPosition: Int, childPosition: Int): Any {
        return childPosition
    }

    override fun getGroupId(groupPosition: Int): Long {
        return groupPosition.toLong()
    }

    override fun getChildId(groupPosition: Int, childPosition: Int): Long {
        return childPosition.toLong()
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    override fun getGroupView(
        groupPosition: Int,
        isExpanded: Boolean,
        convertView: View,
        parent: ViewGroup
    ): View? {
        val item = topCells[groupPosition]
        when (item.type) {
            FilterModelItem.SECTION_TITLE -> {
                val sectionView = View.inflate(context, R.layout.cell_section_title, null)
                val title = sectionView.findViewById<View>(R.id.section_title) as TextView
                title.text = item.title
                return sectionView
            }
            FilterModelItem.SIMPLE_CELL -> {
                val view = View.inflate(context, R.layout.cell_simple, null)
                SimpleCell.fillWithTitle(
                    view,
                    item.title  ?: "",
                    if (item.isChecked) CellAccessoryType.CHECKED else CellAccessoryType.HIDDEN,
                    1
                )
                return view
            }
            FilterModelItem.CATEGORY_CELL -> {
                val categoryView = View.inflate(context, R.layout.cell_simple, null)
                SimpleCell.fillWithTitle(
                    categoryView,
                    item.title  ?: "",
                    if (isExpanded) CellAccessoryType.TOP else CellAccessoryType.BOTTOM,
                    1
                )
                return categoryView
            }
        }
        return null
    }

    override fun getChildView(
        groupPosition: Int,
        childPosition: Int,
        isLastChild: Boolean,
        convertView: View,
        parent: ViewGroup
    ): View {
        val view = View.inflate(context, R.layout.cell_models_list, null)
        val imageView = view.findViewById<View>(R.id.cell_image) as ImageView
        val modelTitle = view.findViewById<View>(R.id.model_title) as TextView
        val years = view.findViewById<View>(R.id.model_years) as TextView
        val arrowImage = view.findViewById<View>(R.id.arrowView) as ImageView
        val model = topCells[groupPosition].models?.get(childPosition)
        var modelChecked = false
        val selectedModel = MyMotApplication.searchManager.model
        if (selectedModel != null) {
            if (selectedModel.id == model?.id ?: 0) {
                modelChecked = true
            }
        }
        if (modelChecked) {
            arrowImage.setImageResource(R.drawable.ic_checked)
            arrowImage.visibility = View.VISIBLE
        } else {
            arrowImage.visibility = View.GONE
        }
        Picasso.get().load(Const.DOMAIN + model?.preview_picture)
            .placeholder(R.drawable.ic_placeholder).into(imageView)
        modelTitle.text = model?.name
        years.text = model?.years
        return view
    }

    override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
        return true
    }
}